probe-rs/src/debug/source_instructions.rs

```diff
@@ -23,6 +24,223 @@ pub struct VerifiedBreakpoint {
+                continue;
+            };
+            // Keep track of the matching file index to avoid having to lookup and match the full path
+            // for every row in the program line sequence.
+            let mut matching_file_index = None;
+            if line_program
+                .header()
+                .file_names()
+                .iter()
+                .enumerate()
+                .any(|(file_index, _)| {
+                    debug_info
+                        .get_path(&program_unit.unit, file_index as u64)
+                        .map(|combined_path: TypedPathBuf| {
+                            if canonical_path_eq(path, &combined_path) {
+                                matching_file_index = Some(file_index as u64);
+                                true
+                            } else {
+                                false
+                            }
+                        })
+                        .unwrap_or(false)
+                })
+            {
+                let Ok((complete_line_program, line_sequences)) = line_program.clone().sequences()
+                else {
+                    continue;
+                };
+                for line_sequence in line_sequences {
+                    let instruction_sequence = InstructionSequence::from_line_sequence(
+                        debug_info,
+                        program_unit,
+                        complete_line_program.clone(),
+                        &line_sequence,
+                    );
+
+                    // Cycle through various degrees of matching, to find the most relevant source location.
+                    if let Some(verified_breakpoint) = match_file_line_column(
+                        &instruction_sequence,
+                        matching_file_index,
+                        line,
+                        column,
+                        debug_info,
+                        program_unit,
+                    )
+                    .or_else(|| {
+                        match_file_line_first_available_column(
+                            &instruction_sequence,
+                            matching_file_index,
+                            line,
+                            debug_info,
+                            program_unit,
+                        )
+                    }) {
+                        return Ok(verified_breakpoint);
+                    }
+                }
+            }
+        }
+        // If we get here, we have not found a valid breakpoint location.
+        Err(DebugError::Other(anyhow::anyhow!("No valid breakpoint information found for file: {}, line: {line:?}, column: {column:?}", path.to_path().display())))
+    }
+}
+
+/// Find the valid halt instruction location that is equal to, or greater than, the address.
+fn match_address(
+    instruction_sequence: &InstructionSequence<'_>,
+    address: u64,
+    debug_info: &DebugInfo,
+) -> Option<Result<VerifiedBreakpoint, DebugError>> {
```

Comment:
Do you expect this to return an error in the future?

Reply:
Yes, I do, and `None` (not returning any results) would be different to one of the "gimli" calls that can `Err`.